powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/common/KeyDerivationService.java

```diff
@@ -0,0 +1,70 @@
+package io.getlime.security.powerauth.app.server.service.behavior.common;
+
+import com.google.common.io.BaseEncoding;
+import io.getlime.security.powerauth.app.server.converter.v3.ServerPrivateKeyConverter;
+import io.getlime.security.powerauth.app.server.database.model.KeyEncryptionMode;
+import io.getlime.security.powerauth.app.server.database.model.entity.ActivationRecordEntity;
+import io.getlime.security.powerauth.app.server.service.exceptions.GenericServiceException;
+import io.getlime.security.powerauth.crypto.lib.config.PowerAuthConfiguration;
+import io.getlime.security.powerauth.crypto.server.keyfactory.PowerAuthServerKeyFactory;
+import io.getlime.security.powerauth.provider.CryptoProviderUtil;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import javax.crypto.SecretKey;
+import java.security.InvalidKeyException;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.spec.InvalidKeySpecException;
+
+/**
+ * Service used for derivation of keys, shared by behaviors.
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+@Service
```

Comment:
Man, I don't like this service on so many levels...

- It accepts DB entity and returns raw key data.
- It autowires one property and creates all other without DI.
- It mixes concerns randomly (derive key, decrypt private key, ...), but at the same time, it only introduces only one new method that could easily be a private method somewhere or a utility method in crypto...

Reply:
Let's have a look tomorrow.

Reply:
We agreed to split the concerns and keep decryption of server private key in behavior classes.